### PR TITLE
CloudWatch: Update query batching logic

### LINKED ISSUE #74563: Cloudwatch Metric Insights Queries: Update logic for grouping queries
Cloudwatch API has a limit of **one** Metric Insights* query per request, otherwise an error ```Metric request error: "ValidationError: Maximum number of queries (1) exceeded\n\tstatus code: 400``` is returned from the API. However, users can send Metric Insights queries combined with one or more Metric Math queries, which doesn't count into the limit.

Screenshot 2023-09-07 at 7 39 05 PM

Currently, all queries are separated in the alerting/expressions engine and sent separately to CW Plugin's QueryData. This means users can add multiple Metric Insights queries in an alert rule (or dashboard panel + expressions) and the queries resolve, without the error above. 
However, if users add a Cloudwatch Math Expression query that references another query in the panel (or alert rule), because of the separation described above, these queries are sent separately and therefore the math expression loses the query it references. 
We would like to support both use cases:
1. Users can add multiple Metric insights queries in an alert or panel. [Example](https://github.com/grafana/support-escalations/issues/7314)
2. Users can add Metric Math queries that reference other queries in the panel/alert, and get relevant data back. [Example 1](https://github.com/grafana/grafana/issues/69873), [Example 2](https://github.com/grafana/grafana/issues/70139)

The approach we decided on is to add logic that will separate Metric queries if there are multiple, but at the same time bundle any queries that need to be bundled (i.e. queries that reference other queries + their referenced queries).

- [Since the limit is only relevant for Metric Insights ](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/cloudwatch-metrics-insights-limits.html)queries, we only need to execute this logic if there are multiple Insights queries in a request. 
- If there are multiple queries, we just need to check if they have any Metric Math queries that reference them, then group the Metric Insights queries with its math query, then send them each in a separate GetMetricData API request.
- There is a [function](https://github.com/grafana/grafana/blob/main/pkg/tsdb/cloudwatch/models/cloudwatch_query.go#L75) that returns query type (math,insights etc.). After this, we need to find the query it references (either the ID from the QueryEditor, or the default ID set [here](https://github.com/grafana/grafana/blob/2fbc75200360ce92f0bd41893ca13953ff9ff743/pkg/tsdb/cloudwatch/models/cloudwatch_query.go#L318C6-L318C6)) and bundle them together.**

**If there are Math queries that reference multiple Metric insights queries, e.g.
```ABS(q_id_1 + q_id_2)``` this is not something we support, as both queries would need to be bundled and sent with the math expression. In this case we would still bundle them, but make sure that we propagate the limit error (`Maximum number of queries (1) exceeded`) we receive from AWS. 

This feature should be developed under a Grafana feature toggle so we are able to gradually roll out and revert if users encounter issues.

[There is a feature toggle](https://github.com/grafana/grafana/pull/74551) `sseGroupByDatasource` that combines the queries in the expressions/alerts engine, we should develop this feature with it enabled. The reason for this is that we need to have all the CW queries reach our backend together, so we can bundle them if needed. 
Once this feature is GA, we should review sseGroupByDatasource toggle and inform grafanaObservabilityMetricsSquad to release it GA (if there aren't issues with other datasources in the meantime) or remove it.

- [ ] Currently, users that send multiple Metrics Insights queries get an error. Implementing this feature will mean that this no longer happens, we should add a remark in the changelog about this. We assume users take responsibility for querying possibly large amounts of data whenever they build a query in Grafana. However, [AWS has a limit on their end](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/cloudwatch-metrics-insights-limits.html) to make sure the impact is limited.

### PR DESCRIPTION
**Pull request by iwysiu**
**What is this feature?**

Batches the CloudWatch metrics queries if needed. Due to CloudWatch API limits, Metrics Insight queries need to be run in separate batches and metric math expression queries need to be run with with any queries they reference.

**Why do we need this feature?**

Currently, all metric queries from a dashboard (without expressions) are sent together and all queries from alerting/expressions are separated.
This means that multiple metric insights queries can be combined in an alert rule, but not in an expressionless dashboard. This also means that users can use metric math expressions that reference other panel queries, but they can't use these queries in alerting or with grafana expressions. A [feature flag](https://github.com/grafana/grafana/pull/74551) has been created to send queries together from the alerting/expressions engine, but we need to improve the batching logic to maintain the multiple metric insights functionality.
This pr fixes the batching logic so that both cases work in panels and alerts/expressions.

**Who is this feature for?**

This is for users who want to use CloudWatch metrics in the above situations.

**Which issue(s) does this PR fix?**:

Fixes #74563

**Special notes for your reviewer:**
I feel like this should have a feature flag (because it messes with the queries) but I'm not sure if it should be public preview (off but open for use) or general availability (on by default)

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Review by kevinwcyu — APPROVED** [review]

**Comment by fridgepoet** [comment]
Hi @iwysiu! I'm curious how it is expected for a metric math referencing multiple metric insight queries to be batched.

**Review by iwysiu — COMMENTED** [review]
> How it is expected for a metric math referencing multiple metric insight queries to be batched?

The idea from the original issue is that both metric insights would be added and then we would pass along the error from CloudWatch. Added a test for it.

File: pkg/tsdb/cloudwatch/get_metric_query_batches_test.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+		Id:               "m4",
+	}
+	math5 := models.CloudWatchQuery{
+		MetricQueryType:  models.MetricQueryTypeSearch,
+		MetricEditorMode: models.MetricEditorModeRaw,
+		Expression:       "PERIOD(i1) * RATE(i3)",
+		Id:               "m5",
+	}
+
+	t.Run("zero insight queries should not batch", func(t *testing.T) {
```
Comment (fridgepoet) [inline pkg/tsdb/cloudwatch/get_metric_query_batches_test.go:61]: `batch` here may not be clear whether it refers to "grouping together" or "separating" -- maybe we could be a little more explicit which one it is in each test title?

File: pkg/tsdb/cloudwatch/get_metric_query_batches_test.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+	insight3 := models.CloudWatchQuery{
+		MetricQueryType: models.MetricQueryTypeQuery,
+		Id:              "i3",
+	}
+	metricStat := models.CloudWatchQuery{
+		MetricQueryType:  models.MetricQueryTypeSearch,
+		MetricEditorMode: models.MetricEditorModeBuilder,
+		Id:               "s1",
+	}
+	math1 := models.CloudWatchQuery{
```
Comment (fridgepoet) [inline pkg/tsdb/cloudwatch/get_metric_query_batches_test.go:30]: it was a little challenging to read in the tests below which queries these math queries are referencing. maybe the names could be like `m1_ref_i1`, `m3_ref_m1_m2` ? or another solution in camel case?

File: pkg/tsdb/cloudwatch/get_metric_query_batches_test.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+			&insight3,
+			&math1,
+			&math5,
+		}
+
+		result := getMetricQueryBatches(batch, logger)
+		assert.Len(t, result, 3)
+		assert.ElementsMatch(t, []*models.CloudWatchQuery{&insight2}, result[0])
+		assert.ElementsMatch(t, []*models.CloudWatchQuery{&insight1, &math1}, result[1])
+		assert.ElementsMatch(t, []*models.CloudWatchQuery{&insight1, &insight3, &math5}, result[2])
```
Comment (fridgepoet) [inline pkg/tsdb/cloudwatch/get_metric_query_batches_test.go:144]: should we comment that this third batch is expected to return an error from AWS?

File: pkg/tsdb/cloudwatch/get_metric_query_batches.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/tsdb/cloudwatch/models"
+)
+
+// nonWordRegex is for spliting the expressions to just functions and ids
+var nonWordRegex = regexp.MustCompile(`\W+`)
+
+// getMetricQueryBatches separates queries into batches if necessary. Metric Insight queries cannot run together, and math expressions must be run
+// with all the queries they reference.
+func getMetricQueryBatches(queries []*models.CloudWatchQuery, logger log.Logger) [][]*models.CloudWatchQuery {
```
Comment (fridgepoet) [inline pkg/tsdb/cloudwatch/get_metric_query_batches.go:15]: (noticed that `logger` and `idToIndex` below are unused)

**Review by fridgepoet — APPROVED** [review]

File: pkg/tsdb/cloudwatch/get_metric_query_batches_test.go (hunk 2 above)
Comment (iwysiu) [reply pkg/tsdb/cloudwatch/get_metric_query_batches_test.go]: done. Camel case is a more go style but I think the underscores are easier to read in this case

File: pkg/tsdb/cloudwatch/get_metric_query_batches.go (hunk 4 above)
Comment (iwysiu) [reply pkg/tsdb/cloudwatch/get_metric_query_batches.go:15]: fixed

**Comment by fridgepoet** [comment]
Linking related issue: https://github.com/grafana/grafana/issues/70139

**Pull request merged**